compiler/src/fuzzer/mod.rs

```diff
@@ -1,194 +1,93 @@
-                        match message {
-                            Value::Text(message) => message,
-                            other => format!("{}", other),
-                        },
-                    );
-                    log::error!("This was the stack trace:");
-                    let db = db.lock().await;
+                vm::Status::Panicked(value) => {
+                    log::error!("VM panicked with value {value}.");
+                    log::error!("This is the stack trace:");
                     vm.tracer.dump_stack_trace(&db, input.clone());
-
-                    let trace = vm.tracer.dump_call_tree();
-                    // PathBuff::new(input.to_path().unwrap())
-                    let mut trace_file = input.to_path().unwrap();
-                    trace_file.set_extension("candy.trace");
-                    fs::write(trace_file.clone(), trace).unwrap();
-                    log::info!(
-                        "Trace has been written to `{}`.",
-                        trace_file.as_path().display()
-                    );
-                    continue 'test_different_closures;
+                    return;
                 }
             }
         }
-        log::debug!("Couldn't find any issues with this function.");
-    }
-}
+    };
 
-async fn test_closure_with_args(
-    db: Arc<Mutex<Database>>,
-    closure_id: &hir::Id,
-    vm: &mut Vm,
-    closure_address: usize,
-    arguments: Vec<Value>,
-) -> TestResult {
-    let closure = vm.heap.export_without_dropping(closure_address);
+    let TearDownResult {
+        fuzzable_closures, ..
+    } = vm.tear_down_module_closure_execution();
 
-    {
-        let db = db.lock().await;
-        println!("Starting closure {closure}.");
-        let use_provider = DbUseProvider { db: &db };
-        vm.set_up_closure_execution(&use_provider, closure, arguments);
-        vm.run(&use_provider, 1000);
-    }
+    log::info!(
+        "Now, the fuzzing begins. So far, we have {} closures to fuzz.",
+        fuzzable_closures.len()
+    );
 
-    match vm.status() {
-        Status::Running => TestResult::StillRunning,
-        Status::Done => TestResult::NoPanic,
-        Status::Panicked(message) => {
-            // If a needs directly inside the tested closure was
-            // dissatisfied, then the panic is not the fault of the code
-            // inside the code, but of the caller.
-            let db = db.lock().await;
-            let is_our_fault =
-                did_need_in_closure_cause_panic(&db, &closure_id, vm.tracer.log().last().unwrap());
-            if is_our_fault {
-                TestResult::WrongInputs
-            } else {
-                TestResult::InternalPanic(message)
-            }
+    for (id, closure) in fuzzable_closures {
+        let mut fuzzer = Fuzzer::new(db, closure.clone(), id.clone());
+        for _ in 0..20 {
+            fuzzer.run(db, 100);
         }
```

Comment:
Shouldn't this accomplish the same?

```suggestion
        fuzzer.run(db, 2000);
```

Reply:
Not quite because if it's fuzzing some input and the function returns normally after 90 instructions, we stop. With the for-loop, we would then try again with a different input etc.

Should we change the name of the function or the behavior of the function to continue fuzzing with a reduced number of instructions left?

Reply:
Yes, I think that would make sense. I lean towards the latter, though we could also let it return how many instructions were actually executed or how many are left

Reply:
Okay, I implemented that. I also thought about this pattern of giving in the number of instructions, which we now use both for the VM directly as well as the fuzzer.

In the long term, instead of providing the number of instructions to run, it could make sense to instead provide a closure that decides whether to keep going. The closure could be called before every instruction.
That would allow us to implement several abortion strategies, e.g. aborting after a certain wall time or after a number of instructions or as our CPU usage rises above some threshold, all without requiring us to keep track of the concrete metric in the code.